File: coding-challenges/sources/leetcode/144. Binary Tree Preorder Traversal/144-BinaryTreePreorderTraversal.swift


fileprivate class TreeNode {
    var val: Int
    var left: TreeNode?
    var right: TreeNode?

    init() {
        self.val = 0
        self.left = nil
        self.right = nil
    }

    init(_ val: Int) {
        self.val = val
        self.left = nil
        self.right = nil
    }

    init(_ val: Int,
         _ left: TreeNode?,
         _ right: TreeNode?) {
        self.val = val
        self.left = left
        self.right = right
    }
 }

fileprivate func preorderTraversal(root: TreeNode?) -> [Int] {
    // base case: if the root is `nil`, return an empty array
    guard let node = root else { return [] }

    // visit the root node
    var preorder = [node.val]
    // visit the left subtree
    preorder += preorderTraversal(root: node.left)
    // visit the right subtree
    preorder += preorderTraversal(root: node.right)

    return preorder
}

import XCTest

final class PreorderTraversalTests: XCTestCase {
    func test() {
        let root1 = TreeNode(1)
        root1.right = TreeNode(2)
        root1.right?.left = TreeNode(3)
        XCTAssertEqual(preorderTraversal(root: root1), [1,2,3])

        XCTAssertEqual(preorderTraversal(root: nil), [])

        let root2 = TreeNode(1)
        XCTAssertEqual(preorderTraversal(root: root2), [1])
    }
}

File: coding-challenges/sources/leetcode/144. Binary Tree Preorder Traversal/144-BinaryTreePreorderTraversal.kt
fun main() {
    val root1 = TreeNode(1)
    root1.right = TreeNode(2)
    root1.right?.left = TreeNode(3)
    println(preorderTraversal(root1) == listOf(1, 2, 3))
    
    println(preorderTraversal(null).isEmpty())
    
    val root2 = TreeNode(1)
    println(preorderTraversal(root2) == listOf(1))
}

class TreeNode(var `val`: Int) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}

fun preorderTraversal(root: TreeNode?): List<Int> {
    if (root == null) {
        return emptyList()
    }
    
    // visit the root node
    val preorder = mutableListOf(root.`val`)
    // visit the left subtree
    preorder.addAll(preorderTraversal(root.left))
    // visit the right subtree
    preorder.addAll(preorderTraversal(root.right))
    
    return preorder
}